superset/sql_parse.py

```diff
@@ -458,3 +460,178 @@ def validate_filter_clause(clause: str) -> None:
                 )
     if open_parens > 0:
         raise QueryClauseValidationException("Unclosed parenthesis in filter clause")
+
+
+def has_table_query(statement: Statement) -> bool:
+    """
+    Return if a stament has a query reading from a table.
+
+        >>> has_table_query(sqlparse.parse("COUNT(*)")[0])
+        False
+        >>> has_table_query(sqlparse.parse("SELECT * FROM table")[0])
+        True
+
+    Note that queries reading from constant values return false:
+
+        >>> has_table_query(sqlparse.parse("SELECT * FROM (SELECT 1)")[0])
+        False
+
+    """
+    seen_source = False
+    tokens = statement.tokens[:]
+    while tokens:
+        token = tokens.pop(0)
+        if isinstance(token, TokenList):
+            tokens.extend(token.tokens)
+
+        if token.ttype == Keyword and token.value.lower() in ("from", "join"):
+            seen_source = True
+        elif seen_source and (
```

Comment:
The challenge here is there's no strong connection to ensure that the consecutive (or near consecutive) tokens are those which are being identified here. I guess the question is how robust do we want this logic. The proposed solution may well we suffice.

The _correct_ way of doing this is more of a tree traversal (as opposed to a flattened list) where one checks the next token (which could be a group) from the `FROM` or `JOIN` keyword and iterate from there.

My sense is that can likely be addressed later. We probably need to cleanup the `sqlparse` logic to junk it completely in favor of something else given that it seems like the package is somewhat on life support.

Reply:
Yeah, in the `insert_rls` function I had to implement tree traversal to get it right. Let me give it a try rewriting this one.

Reply:
@john-bodley I reimplemented it following the same logic as `insert_rls` (recursive tree traversal instead of flattening).